### ISSUE TITLE
Mixing CMake and ndkBuild in a .gradle or multi-project build is not supported

### ISSUE THREAD

**Issue by mcclure** [android/ndk#863]
This is part a documentation bug and part a feature request. It simultaneously touches on ndk, gradle, and the Android Studio android plugin, so I don't know where to file. This page's contributing guidelines say to file at the Android tracker but https://developer.android.com/ndk/guides/cmake says to file here. I have filed an identical bug in the Android issue tracker at https://issuetracker.google.com/issues/120166563 .

## THE PROBLEM

Gradle offers two options for building C code for use with Android NDK: ndkBuild and cmake. They cannot be mixed.

If your .gradle contains both a externalNativeBuild { ndkBuild { path '...' } } } and a externalNativeBuild { cmake { path '...' } } }, it prints the error "A problem occurred configuring project ':appname'. > More than one externalNativeBuild path specified". (I found this error wording unhelpful; when I first encountered it I could not decipher it because (1) the line numbers of neither "path" directive are provided, and (2) due to includes, the two conflicting "path" directives were in different files.)

On paper, Gradle allows for multi-project builds wherein some subprojects are ndkBuild and some are cmake. However, ndkBuild and CMake are isolated in a way which makes such builds in practice broken. Targets/modules from the CMake world cannot be referenced from the ndkBuild world or vice versa, environment variables (say, include paths) can't be passed between the two. The targetsmodules are the most critical problem because they mean you cannot have a CMake subproject as a dependencies {} from an ndkBuild subproject or vice versa. See "A Repro Case" below: ndkBuild subproject A can list CMake subproject B in dependencies {}, *but B will run after A, not before*, which means that A will fail because the outputs of B have not been created yet. It is my theory that this happens because Gradle is not aware of any files which are an output of B and a input of A (because I have not been given a way to describe these files to Gradle), so it will ignore the dependencies {}.

These limitations are not documented. They are not listed in https://developer.android.com/ndk/guides/cmake , or in https://developer.android.com/studio/projects/add-native-code . In https://developer.android.com/studio/projects/gradle-external-native-builds it *is* mentioned "Each module in your Android project can link to only one CMake or ndk-build script file", but if you read that full paragraph it goes on to describe how to fit multiple CMake files or multiple ndk-build scripts while misleadingly eliding what happens if you need both a CMake and an ndk-build file in one project.

## WHY THIS IS A PROBLEM

There are various libraries which are distributed as source and whose build scripts use either ndkBuild, or CMake. Including both of these in a single project or multi-project build is extremely painful and apparently impossible without compromises. Google is encouraging NDK projects to migrate to CMake, but currently using CMake means cutting oneself off from the body of existing ndkBuild libraries.

The example I personally hit: The Oculus Mobile SDK https://developer.oculus.com/downloads/package/oculus-mobile-sdk/ for Oculus Go Android headsets comes with only ndkBuild libraries and sample code; on the other hand LOVR https://lovr.org/ an open-source VR framework I contribute to, has only CMake build scripts. It would be very difficult to port LOVR to ndkBuild because of its size and I do not think I can port the Oculus library to CMake because only the people at Oculus properly understand its internals. I have only been able to build a project with both Lovr and the Oculus libraries by invoking gradle twice (see "unsuccessful workarounds" below), which is very slow, and even this required freakish glue code.

## EXPECTED BEHAVIOR

First off, the behavior should be documented. If mixing CMake and ndkBuild is not supported, that fact should be findable in the documentation. If there's some recommended workaround, such as a certain way of configuring multi-project builds, this technique should be documented.

It should be possible to make a multi-project build where some subprojects are ndkBuild and some are CMake. dependencies {} should work properly in these cases, and I should not have to resort to directly listing the paths of intermediate build directories such as build/intermediates/cmake. Perhaps you can come up with a better way of achieving this than me, but here is how I would make this work: Both ndkBuild and CMake have the notion of a target (CMake calls these "targets", ndkBuild calls them "Modules"). It should be possible to reference an ndkBuild module as a CMake Target and vice versa. I should be able to say "LOCAL_SHARED_LIBRARIES := buildtest" in a Android.mk if my CMakeLists has a "add_library(buildtest ...)", for example. If I have to do some kind of "LOCAL_MODULE := buildtest" "include ${CMAKE_LIBRARY}" to make this work, that's fine.

It would be wonderful if there were some way to pass a variable (a CMake "SET", an ndkBuild "var :=") between the CMake and ndkBuild worlds. I guess this is less critical since variables *can* be bassed down from gradle to both.

It would be nice if the "more than one path specified" error message was clearer.

## A REPRO CASE

Here https://github.com/mcclure/android-ndk is a minimal example of the problem based on the googlesamples android-ndk repository. The repro case is in other-builds/ndkbuild/hello-libs and is a recreation of the actual problem my "LOVR" project encountered in https://github.com/mcclure/lovr-oculus-mobile . The example:

I git clone https://github.com/googlesamples/android-ndk.git . I go in the directory other-builds/ndkbuild/hello-libs . I create a completely minimal CMake library which has one (empty) C source file and a message("BUILDING CMAKE TEST PROJECT") in the CMakeLists. I list it as a dependency of the app/ project in settings.gradle and app/build.gradle. I build this, with:

    (export PATH="/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin":~/Library/Android/sdk/platform-tools:$PATH ANDROID_HOME=~/Library/Android/sdk GRADLE=`pwd`/gradlew; ($GRADLE build))

When I build, I see "BUILDING CMAKE TEST PROJECT" in the flow, so I know that my CMakeLists is running. You can find this version listed as 45ff8b1ce.

I next make my CMake subproject not just a dependency, but I make the app/ project actually depend on it. I amend app/'s Android.mk to have a PREBUILT_SHARED_LIBRARY module. This is *super* ugly and horrific, it requires me to order things very precisely in the Android.mk and directly reference the build/intermediates folder, but this is the only way I know to do it This version is listed as a2a9607ac65. When I try to build a2a9607ac65, something interesting happens. On my first build, it works fine. I then try "git clean -xffd". This erases all build products. If I build after git cleaning, the build fails. It prints:

      Android NDK: ERROR:/Users/mcc/work/gh/other/android-ndk/other-builds/ndkbuild/hello-libs/app/Android.mk:buildtest: LOCAL_SRC_FILES points to a missing file    
      Android NDK: Check that /Users/mcc/work/gh/other/android-ndk/other-builds/ndkbuild/hello-libs/app/../cmakelib/build/intermediates/cmake/debug/obj/armeabi-v7a/libbuildtest.so exists  or that its path is correct   
      /Users/mcc/Library/Android/sdk/ndk-bundle/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.

I then try building the cmakelib subproject standalone, with

    (export PATH="/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin":~/Library/Android/sdk/platform-tools:$PATH ANDROID_HOME=~/Library/Android/sdk GRADLE=`pwd`/gradlew; (cd cmakelib && $GRADLE build))

Once I have done this standalone build of the subproject, I am able to build the top level project again.

In other words: The top level build, which depends on the output of cmakelib, is able to build as long as the cmakelib build runs first, either because it ran on a previous run or I built cmakelib separately. Listing :cmakelib as a "dependency" does not cause it to run first, I actually have to invoke gradle twice. I find invoking gradle twice to be an unacceptably scary (and slow!) solution, in addition to which the PREBUILT_SHARED_LIBRARY build code having to reference build/intermediates is unacceptably scary. This overall-unacceptably-scary solution is, as mentioned, the one currently being used in my "lovr-oculus-mobile" project linked above, which is currently shipping, and the README simply asks the user to invoke gradle twice. There appears to be no other way to do it.

## CONFIGURATION

Gradle version: If I run gradlew -v I get:

    ------------------------------------------------------------
    Gradle 4.4
    ------------------------------------------------------------
    Build time:   2017-12-06 09:05:06 UTC
    Revision:     cf7821a6f79f8e2a598df21780e3ff7ce8db2b82
    Groovy:       2.4.12
    Ant:          Apache Ant(TM) version 1.9.9 compiled on February 2 2017
    JVM:          1.8.0_152-release (JetBrains s.r.o 25.152-b01)
    OS:           Mac OS X 10.13.2 x86_64

Android Plugin Version: My gradle file contains:
	  compileSdkVersion 21
	  buildToolsVersion '27.0.3'
 		  minSdkVersion 21
		  targetSdkVersion 21

Android SDK Tools version:  "Select Tools > Android > SDK Manager, click SDK Tools"
"Android SDK Build-Tools" "Version" is blank and "Status" says "Installed"
"Android SDK Platform-Tools" "Version" is "28.0.1" and "Status" is "Installed"
"Android SDK Tools" "Version" is "26.1.1" and "Status" is "Installed"
NDK" "Version" is "18.1.5063045" and "Status" is "Installed"

**Comment by DanAlbert** [comment]
Thanks for filing this. 100% agree that this would be a good feature to have and that we need to improve the docs to state that it's not currently possible, as well as document workarounds.

> This page's contributing guidelines say to file at the Android tracker but https://developer.android.com/ndk/guides/cmake says to file here. I have filed an identical bug in the Android issue tracker at https://issuetracker.google.com/issues/120166563 .

Yeah, this bug straddles the boundary a bit. Most of what's described here is on the gradle side, so I'll triage and elaborate a bit more on the other bug, but I'm going to leave this open because one of the docs mentioned above is something NDK-specific that we own so I'll get that updated to note this feature gap.

> If there's some recommended workaround, such as a certain way of configuring multi-project builds, this technique should be documented.

If I'm understanding your use case properly, your best bet would be to build the Oculus SDK ahead of time and pull the libraries into your CMake build scripts as [imported targets](https://cmake.org/cmake/help/v3.5/command/add_library.html#imported-libraries). The assumption here is that you're not actively making changes to the Oculus SDK and thus don't need it to rebuild as part of your workflow.

I don't think I actually have understood your use case correctly, since your description of the repro case seems to be the opposite: CMake generates the artifacts that are used by ndk-build. If that's the case then yeah, `PREBUILD_SHARED_LIBRARY` would be my recommended approach.

The change I would make to your approach (again, assuming you only care about rebuilding one of these projects with any frequency) would be to build the dependency ahead of time and check that in somewhere. That gets you out of the mess of having to refer into the intermediates directory.

Obviously this doesn't work when there are bidirectional dependencies between the projects, but I don't think solving that is particularly viable anyway.

**Comment by alexcohn** [comment]
We, in some sense, do run gradle twice to take care of similar dependencies. But we do that by putting the cmake and ndkBuild libraries in separate subprojects (in Android Studio these are known as _modules_ ), and define the dependency between them such that gradle itself builds the two in correct order. Still we have the PREBUILT_SHARED_LIBRARY or IMPORTED library, and we are not very happy to explicitly reference *build/intermediates*.

**Comment by mcclure** [comment]
> and define the dependency between them such that gradle itself builds the two in correct order

@alexcohn, how do you do this? As noted in the bug my biggest problem is that [Gradle is not respecting the ordering of my dependencies when building](https://github.com/mcclure/lovr-oculus-mobile/blob/master/LovrApp/Projects/Android/build.gradle#L3). Maybe there is something I do not understand about Gradle :(

**Comment by alexcohn** [comment]
@mcclure I downloaded your [demo project](https://github.com/mcclure/android-ndk/tree/a2a9607ac65203dfbfac16066607fc3db1b40885/other-builds/ndkbuild/hello-libs), and I must say that your [Android.mk](https://github.com/mcclure/android-ndk/blob/a2a9607ac65203dfbfac16066607fc3db1b40885/other-builds/ndkbuild/hello-libs/app/Android.mk), but it will take some time to fix it to work correctly with Android Studio. I don't think I can fix [lovr-oculus-mobile](https://github.com/mcclure/lovr-oculus-mobile), though:

I believe that even these days, Android Studio is not well suited to handle big C++ projects. Don't forget that many traditionally use autoconf and other tricky build environments, and we often require standalone toolchain to cross-compile for Android.

But even the ndk-build projects that have thousands of CPP files and dosens of modules, simply  Android Studio C++ indexing to crawl and eventually out of memory. I would say that before introducing support for mixing **CMake** and **ndk-build**, Android Studio needs some mechanism that will stop introspection and let us easily import 3rd party libraries for which we don't usually need C++ indices.

**Comment by eddielong** [comment]
100% we suffer immensely from this issue too, studios indexing causes slow downs and out of memory conditions often due to third parties (boost etc.). We don't even need the indexing so if entire folders could be omitted that would eb great. Believe clion can do it

**Comment by enh** [comment]
(file Studio bugs under "C++" here instead: https://source.android.com/setup/contribute/report-bugs)

**Comment by mcclure** [comment]
@alexcohn Thanks for trying. I actually cannot get Android Studio to work with NDK projects on my computer, so I build from the command line always.

**Comment by alexcohn** [comment]
I pushed the simple fix to GitHub.

The problem that you encountered is that before build strats, Android Studio must synchronize the modules. For that, it runs **ndk-build** with _dry-run_ parameter. The PREBUILT_SHARED_LIBRARY is very strict, and it crashes the build because the SRC file is not preset (yet), even though the file is not needed during this synchronization step. The actual build does respect dependency order, at least in my experience.

In the proposed [pull request](https://github.com/mcclure/android-ndk/pull/1), I simply replace PREBUILT_SHARED_LIBRARY  with BUILD_SHARED_LIBRARY when libbuildtest.so is not there yet.

As a side note, I have also fixed the way you define LOCAL_SRC_FILES: these are expected to be relative to LOCAL_PATH. I hope this will help you to fix your production scripts.

**Comment by alexcohn** [comment]
> _I actually cannot get Android Studio to work with NDK projects on my computer, so I build from the command line always._

I guess this happens because the size of your C++ code (with all 3rd party libraries) is beyond the capabilities of this IDE. 

One clarification: do you run gradle from command line, or you run ndk-build and use the prebuilt libraries in your Android gradle project?

**Comment by alexcohn** [comment]
Actually, this happens to repeat the issue https://github.com/android-ndk/ndk/issues/676 which may be handled at r20

**Comment by DanAlbert** [comment]
The non-gradle CMake doc does now link to the gradle docs (which are where the explanation of this belongs), so the core NDK work here should be done. Will reopen if the AGP side ends up needing something from us.

**Comment by mcclure** [comment]
Hi, could you clarify, is it now possible to mix CMake and ndkBuild in a single project? How is this done? I'm afraid I don't understand the significance of #676 .

(My project is still being built very awkwardly as two separate gradle invocations because my core project is still using CMake and Oculus still exclusively distributes their sample code as ndkBuild.)

**Comment by DanAlbert** [comment]
No, intermodule native dependencies of any kind are still not supported by the Android Gradle Plugin. It appears to be something that does require a gradle plugin fix rather than anything in the NDK proper though, so I was closing this in favor of https://issuetracker.google.com/issues/120166563. If support for intermodule native dependencies is added, the build systems used shouldn't be an issue.

#676 isn't really related. It's a sanity check in ndk-build that fails if your prebuilts don't exist, which makes it harder to hack around the problem when using ndk-build. CMake will have all the same problems though.

**Comment by mcclure** [comment]
@DanAlbert Thank you for explaining.

**Comment by alexcohn** [comment]
can't this be addressed by prefab? I.e. if module* B depends on AAR that is produced by module A?

*) here I use "module" in terms of Android Studio, which is a "project" in gradle-speak (as opposed to root project).

**Comment by DanAlbert** [comment]
tl;dr Prefab could definitely be the mechanism for this, if intermodule native dependencies were supported. I answered in more detail on the other bug, since it'd benefit from getting the IDE/gradle folks eyes on it too.

> *) here I use "module" in terms of Android Studio, which is a "project" in gradle-speak (as opposed to root project).

Same. That inconsistency is awful :(

**Comment by alexcohn** [comment]
I have forged an experimental project where libraries are passed to app via Prefab: [hello-libs/unified-aar](https://github.com/alexcohn/ndk-samples/tree/prefab/sync/prefab/hello-libs). Here we have CMake on both modules, but this should not matter because Prefab is build-system agnostic.

Unfortunately, this does not resolve the multi-project build cleanly. I still have to build the lib first, and app as a separate step.

I can 'manually' establish dependency of `:app:preBuildDebug` on [`:gen-lib:bundleDebugPrefabAar`](https://github.com/alexcohn/ndk-samples/blob/prefab/sync/prefab/hello-libs/gen-libs/build.gradle#L98), and make sure the AAR is up-to-date when I build the **:app** module, but the updated AAR is not passed to **:app** CMake, so this is pretty useless.

Basically, the missing step is to trigger re-sync of **:app** module programmatically when **:gen-libs:bundleDebugPrefabAar** completes (it could be part of the intermediate **:distribution** module functionality). 

See also https://issuetracker.google.com/issues/120166563#comment8
